Forms events triggers refactoring

gevou opened the pull request.

gevou committed:
replaces Forms.trigger with jQuery $.trigger

gevou committed:
correct way of handling event parameters

gevou committed:
reworked tests to accommodate refactored event handling

gevou committed:
whitespace & commented-out code clean-up

cwohlman commented on lib/forms.js:
```diff
@@ -123,36 +97,27 @@ Forms.change = function (e, tmpl) {
-Forms.submit = function (e, tmpl) {
+Forms.submit = function (e, tmpl, customDoc) {
   // Attach the document, in case the user has some custom handling they
   // want to happen regardless of form validation.
```
This comment can go away

cwohlman commented on lib/forms.js:
```diff
@@ -123,36 +97,27 @@ Forms.change = function (e, tmpl) {
 
-  var doc = tmpl.doc.get();
-  e.doc = doc;
+  var doc = customDoc ? customDoc : tmpl.doc.get();
```
Could you comment as to why you decided to support custom document overrides?

gevou replied:
The intention was to provide flexibility on custom submit actions. e.g. "submit A" button needs to set fields XY and "submit B" button needs to set some values to fields AB. 
One can of course modify the document before submitting it - but in the above way you can do it in one line. More of an easy to implement nice-to-have than a solid use-case-backed feature to be honest.

cwohlman commented on lib/forms.js:
```diff
@@ -123,36 +97,27 @@ Forms.change = function (e, tmpl) {
-  var doc = tmpl.doc.get();
-  e.doc = doc;
+  var doc = customDoc ? customDoc : tmpl.doc.get();
+  // e.doc = doc;
```
This line can go

gevou replied:
ok

cwohlman commented on lib/forms.js:
```diff
@@ -78,44 +78,18 @@ Forms.set = function (doc, name, value) {
   // e.propertyName, e.propertyValue if they exist.
   // Will trigger documentChange event if the document is updated
-Forms.change = function (e, tmpl) {
+Forms.change = function (e, tmpl, change) {
```
Let's rename this property `changes`, for now it's only allowed to have a single property, but in the future we can support updating multiple properties at a time via this paradigm

cwohlman replied:
Alternately you could handle the user passing the key and value as two separate arguments

gevou replied:
ok

cwohlman commented on lib/forms.js:
```diff
@@ -123,36 +97,27 @@ Forms.change = function (e, tmpl) {
-      doc: doc
-      , errors: errors
-    });
+    $(e.currentTarget).trigger('documentInvalid', [ doc, errors ])
```
@ianserlin, @gevou  What's your take on the relative merits of this approach: 
`$(...).trigger('documentInvalid', [ doc, errors ])`
compared with this approach:
`$(...).trigger('documentInvalid', {doc: doc, errors: errors})`

ianserlin replied:
The first is more explicit, the second is more extensible without app developers having to modify existing code. 

So it depends on how likely we believe it will be that more information will be passed this way in the future. If very low, I would prefer the first as an app developer. If we think medium to high, my life as an app developer will be easier via the second approach (`{doc: doc, errors: errors}`) because my handler code won't need to change to accommodate new/re-ordered formal parameters.

gevou replied:
I chose the array approach simply because it felt more "familiar" to the likes of aFunction.apply(contextObj, [argsArray]). So there is no architectural rationale behind this choice.
Beyond that Ian's comment fully reflects my thoughts too. 
At the moment, I cannot think of what additional parameter could be passed along from `documentInvalid`. Maybe the pairs of invalid fields-values? (sounds more like nice-to-have than a feature though)

cwohlman replied:
I agree, I can't see needing to pass more than one or two arguments in any of the Forms event handlers, and those objects fit a fairly predictable pattern. As Ian said, if we think this api is going to change it might be better to use an object with properties, but I don't see us adding arguments to these events in the near future.

cwohlman commented on tests/forms.js:
```diff
@@ -130,30 +134,13 @@ Tinytest.add('Forms - documentInvalid event is triggered when form is invalid',
-  div.on('submit', function (e) {
-    test.equal(e.doc, {
+    test.isNotUndefined(doc, 'documentInvalid event: no doc parameter');
+    test.isNotUndefined(errors, 'documentInvalid event: no errors parameter');
```
Nice job using other tinytest methods :)

cwohlman commented on tests/forms.js:
```diff
@@ -282,18 +269,20 @@ Tinytest.add('Forms - Event detection - submit method is called', function (test
   });
 
-  div.find('input').trigger('mockSubmit');
+  div.find('input').trigger('mockSubmit', doc.get().doc);
```
There's no point in passing this document and then testing that it was received on the other end because that doesn't touch any of our code.

ianserlin replied:
But it does prove that it was passed by our code.

gevou replied:
`div.on('documentSubmit', function (e,doc) {...` will only fire after a successful document submit action and `doc` will contain the updated document. This part of the test does check our code.  

`div.on('mockSubmit', function (e, doc) {...` checks that the triggered event does propagate from the `input` element to the parent template. This was added as an intermediate step check while building the test. Strictly speaking, it can be regarded as redundant since `documentSubmit` will fail if `mockSubmit` fails in any case.

cwohlman commented on tests/forms.js:
```diff
@@ -251,8 +238,8 @@ Tinytest.add('Forms - Event detection - change event updates doc', function (tes
-  'mockSubmit': function (e, tmpl) {
-    Forms.submit( e, tmpl );
+  'mockSubmit': function (e, tmpl, doc) {
+    Forms.submit( e, tmpl, doc );
```
I don't think we should support custom documents as part of the submit event... we want, if possible, all document changes to go through the Forms.change handler, because the point of this package is for the form to be reactive. So even if the form is entirely virtual (e.g. there are no related `<form>` tags or `<input>` tags) we still want the app developer to use the propertyChange event and then depend on the reactive doc.

gevou replied:
ok - I'm updating the pull request

gevou committed:
Renames `change` attribute to `changes` to reflect actual usage intention

gevou committed:
comment cleanup

gevou committed:
Removes custom document attribute for `submit` event and method.
Any changes to the document should take place before the submit action or method is triggered.

cwohlman commented:
This PR is now obsolete
